### ISSUE TITLE
Use one package library per package?

### ISSUE THREAD

**Issue by krlmlr** [r-lib/revdepcheck#128]
This gives both isolation and performance: Each package is installed only once, into its dedicated package library. Each worker then uses a unique `.libPaths()`, which may be very long.

I'm tempted to start a new package "multilib" for the purpose of managing such split package libraries, but I wanted to double-check first if this is the scope of pkgman or perhaps another package I'm missing.  The "multilib" package would offer replacements for `library()` and `require()` that temporarily change `.libPaths()` so that all dependencies can be found, and also an entry point to compute library paths so that a particular package with all of its declared dependencies is available. It would also need an `install()` entry point.

## Performance test

I've made a few quick checks on my local package library of ~800 packages. First, I created 800 package libraries, all in `~/R-separate`, where each corresponds to one package:

```sh
ls -d ~/R/x86_64-pc-linux-gnu-library/3.4/* |
  sed -r 's#^(.*)/([^/]*)$#echo \2; mkdir -p R-separate/\2; ln -s \1/\2 R-separate/\2/\2#' |
  sh
```

(Replace the last `sh` with `head` or `less` to see what's going on before running.)

Then, I tested the performance of `library(tidyverse)` with both my regular and the split library:

### Flat library

``` r
system.time(suppressPackageStartupMessages(library(tidyverse)))
#>    user  system elapsed 
#>   1.272   0.136   1.409
```

### Split library

``` r
.libPaths(dir("~/R-separate/", full.names = TRUE))
system.time(suppressPackageStartupMessages(library(tidyverse)))
#>    user  system elapsed 
#>   1.264   0.156   1.426
```

This seems to scale, at least for 800 packages.

**Comment by gaborcsardi** [comment]
Smart idea, but I am a bit worried that it will freak out some packages or even base R functions.

**Comment by krlmlr** [comment]
Thanks. Can you please be a bit more specific? What problems do you expect?

**Comment by gaborcsardi** [comment]
This is usually a vector with 2-5 entries, 10 is already extreme. Making it a vector of (potentially) thousands of entries puts us in an area that is not just not well tested, but never actually used by anyone.

E.g. does `installed.packages()` work? How fast? How about `sessioninfo::session_info()`?

A more concrete example is `callr` passing `.libPaths()` to `R CMD check` in an environment variable. (Because maybe there is no other portable way?) Length of environment variables is limited on Windows.

**Comment by krlmlr** [comment]
I briefly checked, `installed.packages()` and `session_info()` seem to scale. But I agree that we need more testing in this unknown land.

Can we use [`$R_CHECK_ENVIRON`](https://cran.r-project.org/doc/manuals/r-devel/R-exts.html#Checking-and-building-packages) to set the library path?

Another option would be to just use a library full of symlinks/junction points, but I remember you weren't sure about this approach either?

I really like the isolated package libraries we have now, but I'm also looking for a way to avoid repeated installation of packages.

**Comment by gaborcsardi** [comment]
> Can we use $R_CHECK_ENVIRON to set the library path?

Yeah, we can, but that's still only `R CMD check`, ideally all of `R CMD` should work. But I guess we can just use `R_ENVIRON` for these.

> Another option would be to just use a library full of symlinks/junction points, but I remember you weren't sure about this approach either?

What's the situation of symlinks of oldish windows? Like windows 2008? We don't use windows currently, and `rcmdcheck` in its current form probably does not even work on windows, but eventually it should.

>I really like the isolated package libraries we have now, but I'm also looking for a way to avoid repeated installation of packages.

Yeah, let's keep thinking about this.

**Comment by gaborcsardi** [comment]
> > Another option would be to just use a library full of symlinks/junction points, but I remember you weren't sure about this approach either?

> What's the situation of symlinks on oldish windows? Like windows 2008? We don't use windows currently, and rcmdcheck in its current form probably does not even work on windows, but eventually it should.

But even if symlinks are problematic on windows, we could probably use them on other platforms.

**Comment by gaborcsardi** [comment]
Btw. a related problem is build time dependencies. E.g. when you revdep-check `R6`, and it is a build-time dependency of package `foo`, then technically `foo` would need be re-installed for the new version of `R6`.

But of course we are not doing this with the current setup either.....

**Comment by krlmlr** [comment]
You mean Windows 98? Junction points work on NTFS 3.0, the default FS in Windows 2000.

If we have one "master" library we create symlinks to, we can also have two, one using "old" and another using "new" as starting point. Would that give full build-time-dep support? Maybe we also can check the upgrade path: installing "new" into a previously "old" package library.

What happens if I run `crancache::install_packages()` using 24 workers and 24 different packages with the same target library?

**Comment by gaborcsardi** [comment]
> You mean Windows 98? Junction points work on NTFS 3.0, the default FS in Windows 2000.

No, I meant win-builder. But then that should be OK, assuming NTFS.

> What happens if I run crancache::install_packages() using 24 workers and 24 different packages with the same target library?

In theory it should be OK, except when multiple processes are installing the same package. Then `install.packages()` locks without timeout AFAIR, so some installations will fail.

**Comment by gaborcsardi** [comment]
Btw. why is BH 700 MiB for you? It is ~138 MiB for me. Also, installing it from binary is 3-4 secs: 
```
❯ system.time(crancache::install_packages("BH", lib="/tmp/lib"))

The downloaded binary packages are in
	/var/folders/59/0gkmw1yj2w7bf2dfc3jznv5w0000gn/T//RtmpFmhmoh/downloaded_packages
   user  system elapsed
   1.59    1.61    3.28
```
If you install it 200+ times, then this is painful, and I guess it is also slower if the system is doing a lot of IO from 50+ check processes...

**Comment by krlmlr** [comment]
I don't remember where I said BH was 700 MiB. My system says it's 142M. It still has >10000 files, which may be a problem too.

**Comment by gaborcsardi** [comment]
Right, you did not say that, I misrememered: https://github.com/r-lib/revdepcheck/pull/53#issuecomment-314877519

**Comment by hadley** [comment]
I'd say this is out of scope for revdepcheck
